### PR TITLE
[webui] customized trial

### PR DESCRIPTION
**Pull request by lvybriage**

### CODE REVIEW

File: src/webui/src/components/trial-detail/TableList.tsx [hunk 1]
```diff
@@ -458,7 +502,14 @@ class TableList extends React.Component<TableListProps, TableListState> {
                         className="titleColumn"
                     />
                 </Modal>
+                {/* compare trials based message */}
                 <Compare compareRows={selectRows} visible={isShowCompareModal} cancelFunc={this.hideCompareModal} />
+                {/* clone trial parameters and could submit a customized trial */}
+                <Customize
+                    visible={isShowCustomizedModal}
```
Comment (liuzhe-lz) [inline src/webui/src/components/trial-detail/TableList.tsx]: Suggest move this out. Only render this element when it's visible.

File: src/webui/src/components/Modal/CustomizedTrial.tsx [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+    }
+
+    render() {
+        const { visible, closeCustomizeModal, hyperParameter } = this.props;
+        const { isShowSubmitSucceed, isShowSubmitFailed, isShowWarning, customID } = this.state;
+        const {
+            form: { getFieldDecorator },
+            // form: { getFieldDecorator, getFieldValue },
+        } = this.props;
+        const parameters = (hyperParameter !== '') ? JSON.parse(hyperParameter) : '';
```
Comment (liuzhe-lz) [inline src/webui/src/components/Modal/CustomizedTrial.tsx]: If your purpose is to mark empty hyper-parameters, `null` is better than `''`.

File: src/webui/src/components/trial-detail/TableList.tsx [hunk 3]
```diff
@@ -458,7 +502,14 @@ class TableList extends React.Component<TableListProps, TableListState> {
...
+                    hyperParameter={trialParameter}
```
Comment (liuzhe-lz) [inline src/webui/src/components/trial-detail/TableList.tsx]: I think it's good to pass as few props as possible. `Customize` only needs to know which trial (ID) it's copying.

File: src/webui/src/components/Modal/CustomizedTrial.tsx [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+    closeFailedHint = () => {
+        // also close customized trial modal
+        if (this._isCustomizeMount === true) {
+            this.setState(() => ({ isShowSubmitFailed: false }));
+            this.props.closeCustomizeModal();
+        }
+    }
+
+    componentDidMount() {
+        this._isCustomizeMount = true;
```
Comment (liuzhe-lz) [inline src/webui/src/components/Modal/CustomizedTrial.tsx]: Official react doc says we should not check "is mounted" to update states.
Instead we should cancel handlers on unmount.
If there is no handler, than this check does nothing.

Comment (lvybriage) [reply src/webui/src/components/Modal/CustomizedTrial.tsx]: OK.

File: src/webui/src/components/Modal/CustomizedTrial.tsx (hunk 2 above)
Comment (lvybriage) [reply src/webui/src/components/Modal/CustomizedTrial.tsx]: Ok.

File: src/webui/src/components/trial-detail/TableList.tsx (hunk 1 above)
Comment (lvybriage) [reply src/webui/src/components/trial-detail/TableList.tsx]: Ok.

File: src/webui/src/components/trial-detail/TableList.tsx (hunk 3 above)
Comment (lvybriage) [reply src/webui/src/components/trial-detail/TableList.tsx]: Ok.

File: src/webui/src/components/Modal/CustomizedTrial.tsx [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+                    destroyOnClose={true}
+                    maskClosable={false}
+                    centered={true}
+                >
+                    {/* search space form */}
+                    <Row className="hyper-box">
+                        <Form>
+                            {
+                                Object.keys(parameters).map(item => {
+                                    return (
```
Comment (liuzhe-lz) [inline src/webui/src/components/Modal/CustomizedTrial.tsx]: `{ return ( blahblah ) }` is equal to `blahblah`

File: src/webui/src/components/trial-detail/TableList.tsx (hunk 1 above)
Comment (lvybriage) [reply src/webui/src/components/trial-detail/TableList.tsx]: > Suggest move this out. Only render this element when it's visible.

We can't see the `succeed modal` if `<Customize>` component's status is hidden. So...

File: src/webui/src/components/Modal/CustomizedTrial.tsx [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+                        || customized[item] > searchSpace[item]._value[1]) {
+                        flag = true;
+                        return;
+                    }
+                }
+            }
+        });
+        if (flag !== false) {
+            // open the warning modal
+            this.setState(() => ({ isShowWarning: true, customParameters: customized }));
```
Comment (liuzhe-lz) [inline src/webui/src/components/Modal/CustomizedTrial.tsx]: `() =>` is unnecessary in all `setState`s.
`this.setState(() => ({ k: v }));` is effectively identical to `this.setState({ k: v });`.

**Comment by chicm-ms** [comment]
Merged into PR 1695

**Pull request closed without merging**